### PR TITLE
Flowcontrol scheduler

### PR DESCRIPTION
**Pull request by bkutil**
This is a framework for future flow control scheduler. It is limited in several ways (which will be adressed in future PRs):

### CODE REVIEW

**Comment by bkutil** [comment]
Depends on https://github.com/juttle/juttle/pull/522

**Comment by bkutil** [comment]
cc: @demmer @davidvgalbraith @dmajda 

I'll create/update a tracking issue with missing parts (as a checklist), then start working through them.

**Comment by bkutil** [comment]
refs: #167

**Comment by bkutil** [comment]
Rebased.

@demmer @davidvgalbraith bump. The state of the PR is equivalent to what we already have. And additional features of the flow control (see the checklist in #167) can be added gradually.

**Comment by demmer** [comment]
@bkutil As we discussed, I added a high-level design writeup of the approach I had in mind to the top of the flow-control.js file. This represents what we talked about in person so please take a look and adjust / update.

**Comment by bkutil** [comment]
@demmer I removed the logic of driving the flowcontrol via `setInterval`, could you please take a look?

I'm not sure about the correctness of ticks in the current implementation - they're driven by `enqueue`, while I think they should be independent of how often the source calls back (if their purpose is to act as a heartbeat and move the window forward while source is either lagging or silent for a long time).

Also, it is not clear how ticks will interact with emitting points when sources will be synchronized and emitting can be blocked until a read fills back the queue. But this is a separate problem to think about later.

File: lib/runtime/flow-control.js [hunk 1]
```diff
@@ -1,5 +1,53 @@
...
+    to occur.
+
+    To accomplish this, the scheduler synchronizes the time range that's being
+    read from each source, potentially stalling faster sources in order for
+    slower sources to catch up. All processing in the middle of the flowgraph
+    occurs without interactions with the scheduler. To handle the case where a
+    source is faster than a sink, each sink is responsible for tracking the
+    amount of outstanding data it has buffered. If it exceeds a threshold, then
+    the sink signals the flow control scheduler to stop reading from all sources
+    and restarts things once the sink drainks.
```
Comment (davidvgalbraith) [inline lib/runtime/flow-control.js]: sink drainks

File: lib/runtime/flow-control.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        this.maxQueueSize = this.options.maxQueueSize || 100000;
+        this.defaultQueueSize = this.options.defaultQueueSize || 10000;
+
+        this.logger_name = 'flow-control';
+        this.logger = JuttleLogger.getLogger(this.logger_name);
+    }
+
+    register(proc) {
+        var pname = proc.params.pname;
+        if (!this.procs[pname]) { this.procs[pname] = proc; }
```
Comment (davidvgalbraith) [inline lib/runtime/flow-control.js]: `else` should we throw an error or is it OK?

**Pull request closed without merging**